### PR TITLE
Adding ML post

### PR DESCRIPTION
**Pull request by patriciorocca**

### CODE REVIEW

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+In order to estimate conversion rates, we implemented a SGD (Stochastic
+Gradient Descent) algorithm or, more specifically, [a FTRL-P][ftrlp] (Follow the
+Regularized Leader Proximal) algorithm with [Adagrad][adagrad] (Adaptive
+Learning Rate) and [ElasticNet][enet] (Lasso and Ridge) regularization. This is
+a state-of-the-art algorithm mostly used in online advertising because of its
+[convenient trade-off][sgd] between optimization and estimation errors in
+scenarios where computational capacity ---and not available data--- is the
+real bottleneck. Every day our system is able to learn from an online stream of
+tens of millions of events published by the bidder into a very
```
Comment (jdemonasterio) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: @memeplex what do you think of using `messages` or `auctions` instead of `events`? Even though they are semantically _events_, I think this might add some confusion. I would prefer using `events` for clicks, installs, and in-app-events (Y variable), whilst messages/auctions for the estimator's input (X variable).

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+$$
+Of course, the goal of the algorithm is to learn $\theta_t$ from
+a succession of input events $(x_1,y_1), \ldots, (x_t,y_t)$. For this
+we minimize the sum over each previously seen input event of:
+(i) (The gradient of) the log-likelihood loss, which is a [logloss][] for the
+logistic model;
+(ii) A stabilizing term that introduces some strong convexity into the mix;
+(iii) Regularizing terms that combine lasso and ridge regularization.
+
+After filling in the details and translating everything to Greek:
```
Comment (jdemonasterio) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: LOL

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+\theta_{t+1} = \underset{\theta}{\operatorname{argmin}}\ 
+{g_{1:t}}^T \theta +
+\frac{1}{2\eta_0}\sum_{s=1}^t ||\theta_s-\theta||^2_{A_s - A_{s-1}} +
+\lambda_1 ||\theta||_1 + \frac{\lambda_2}{2}||\theta||^2_2
+$$
+where:
+* $g_{1:t}$ is the sum of the previous gradients $g_1, \ldots, g_t$ of the
+  loss function.
+* $A_t = (\sum_{s=1}^t g_s g'_s)^{1/2}$, which
+  [minimizes the regret bound][minregret] over Mahalanobis norms for projected
```
Comment (jdemonasterio) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: @memeplex what does the regret bound mean in this context? Is it a ML term?

Comment (jdemonasterio) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: and this expression `||\theta_s-\theta||^2_{A_s - A_{s-1}}`, what does the subindex mean? where the norm is given in the base relative to `A_s - A_{s-1}`?

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+auction is announced and any interested bidder has to answer with a bid price
+within a time constraint of about 100 milliseconds. The bidder that wins the
+auction pays the second-highest price and obtains the right to print a creative
+(which is just jargon for displaying an ad) on a publisher site. If the banner
+is clicked by the user, further goal events might be tracked, v.g. the
+application may be installed or opened. Obtaining a goal event is called a
+conversion.
+
+One main component of our platform is the machine learning subsystem that feeds
+the bidding subsystem with predictive models of conversion rates for different
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: that feeds the bidding system -> ML subsystem is part of a bigger "system", or maybe they are both subsystems, but the repetition is confusing.

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+events of interest (clicks, installs, opens, etc.). A conversion rate for an
+event $E$ is the conditional probability $p(E|C,T)$ of getting that event by
+printing some creative $C$ in the context of the current RTB transaction $T$.
+Why is it so important for our business to have good estimates of these rates?
+Simply because we sell conversions but we pay for impressions. For example,
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: "Simply because we sell conversions but we pay for impressions." Missing comma before but, and some "mystery" could be added with the ":" -> "Simple: we sell conversions, but we pay for impressions."

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+suppose one client pays us $\$1$ per install and we estimate the install
+conversion rate to be around $0.1$ (just cheating... real world figures are well
+under a dismaying $0.001$ `:(`). Then, assuming our creative is going to get
+printed, we should expect an average profit of $\$0.1 - c$, where $c$ is the
+expected cost of the impression. If our rate estimates are grossly miscalculated
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: If our rate estimates are grossly miscalculated, we ...

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+we will be losing money by valuating the expected income too low (thus missing
+opportunities with $c < \$0.1$) or too high (thus buying non-opportunities with
+$c > \$0.1$)[^risk]. The bottom line is that our bidder may be able to pick an
+optimal pair (bid, creative) for the market transaction in course from:
+(i) The estimated conditional conversion rate for the goal event.
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: Semicolons and "." in the final item are better, when the items are not full sentences: [see here](https://english.stackexchange.com/questions/75591/semicolons-at-the-end-of-list-items).

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+Why is it so important for our business to have good estimates of these rates?
+Simply because we sell conversions but we pay for impressions. For example,
+suppose one client pays us $\$1$ per install and we estimate the install
+conversion rate to be around $0.1$ (just cheating... real world figures are well
+under a dismaying $0.001$ `:(`). Then, assuming our creative is going to get
+printed, we should expect an average profit of $\$0.1 - c$, where $c$ is the
+expected cost of the impression. If our rate estimates are grossly miscalculated
+we will be losing money by valuating the expected income too low (thus missing
+opportunities with $c < \$0.1$) or too high (thus buying non-opportunities with
+$c > \$0.1$)[^risk]. The bottom line is that our bidder may be able to pick an
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: "may be able",  "is able" or "might be able"? It depends on the likeliness of that event. If the bidder always picks an optimal pair, then "is able" is correct. If it is unusual -> "might", if usual -> "may".

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+Regularized Leader Proximal) algorithm with [Adagrad][adagrad] (Adaptive
+Learning Rate) and [ElasticNet][enet] (Lasso and Ridge) regularization. This is
+a state-of-the-art algorithm mostly used in online advertising because of its
+[convenient trade-off][sgd] between optimization and estimation errors in
+scenarios where computational capacity ---and not available data--- is the
+real bottleneck. Every day our system is able to learn from an online stream of
+tens of millions of events published by the bidder into a very
+lightweight middleware implemented on top of [ZMQ][zmq].
+
+The first step after receiving an event from the stream is to turn it into a
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: Just a warning: "turn it" could be wrong if the event is later used in a raw format (probably the numerical vector is "generated", "created" or "built from the event", but the event still remains).

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+numerical vector that the estimator can consume. Since most of our variables are
+categorical and since we consider many interactions between these variables, the
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: The second "since" can be removed.

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+output vector will be a sparse binary array of "dummy" variables. We feed this
+vector to a cluster of estimators, each one parameterized according to a grid of
+parameters controlling strength of regularization, number of hashing bits (yes,
+we take advantage of the [hashing trick][trick] as everybody else does),
+estimator memory, etc. Periodically, we evaluate the estimators in the cluster
+using a mix of progressive validation and bootstrap validation in order to
+select the best performing parameterization. The chosen estimator is then
+converted to a tree-like data structure thoroughly optimized for fast model
+evaluation and, finally, it's exported to the bidding subsystem.
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: Should "bidding subsystem" have been replaced by "bidding system" above, it must be replaced here and elsewhere.

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+$\theta$ denote the vector of model coefficients. Then our prediction at time
+$t$ is computed from the logistic regression model:
+$$
+\hat{y_t} = \frac{1}{1 + e^{-{\theta_t}^T x_t}}
+$$
+Of course, the goal of the algorithm is to learn $\theta_t$ from
+a succession of input events $(x_1,y_1), \ldots, (x_t,y_t)$. For this
+we minimize the sum over each previously seen input event of:
+(i) (The gradient of) the log-likelihood loss, which is a [logloss][] for the
+logistic model;
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: lowercase if using the ";" convention for lists (see above)

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+  loss function.
+* $A_t = (\sum_{s=1}^t g_s g'_s)^{1/2}$, which
+  [minimizes the regret bound][minregret] over Mahalanobis norms for projected
+  gradient descent. That said, in real life we just consider the diagonal matrix
+  $diag(A_t)$ for the sake of computation.
+* $\eta_0$ is the initial learning step.
+* $\lambda_1$ is the strength of the lasso regularization.
+* $\lambda_2$ is the strength of the ridge regularization.
+
+Regularization terms aside, this is mostly equivalent to an online gradient
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: Too long, especially the parenthetical sentence.

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+descent formulation with learning rate $\eta_t = \eta_0 A_t^{-1}$, but solving
+the minimization in a follow-the-leader fashion allows for an effective
+implementation of lasso regularization that produces coefficients which are
+exactly zero (and, hence, more sparse models that are cheaper to store and
+transfer and, most importantly, way faster to evaluate). Although at first sight
+it might seem that implementing the follow-the-leader update step ---which sums
+over every previous input--- would be harder than implementing the gradient
+descent update step ---which just cares about the last update--- this first
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: Missing comma before "this first", to separate both subclauses (still a little loaded for a single sentences, with the  "---" comments separating too much the related "it might seem" and "this first impression").

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+{z_t}^T \cdot \theta +
+\frac{1}{2\eta_0} ||\theta||^2_{A_t} +
+\lambda_1 ||\theta||_1 + \frac{\lambda_2}{2}||\theta||^2_2
+$$
+where $z_t = {g_{1:t}} - \frac{1}{\eta_0} \sum_{s=1}^t (A_s - A_{s-1})\theta_s$,
+which can be cheaply calculated in an incremental way as:
+$$
+z_t = z_{t-1} + g_t + \frac{1}{\eta_0} (A_t - A_{t-1})\theta_t
+$$
+Now, when $A_t$ is diagonal the closed form of the solution is just:
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: is diagonal, the closed

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+$$
+\theta_{t,i} =
+\begin{cases}
+0 & \text{if } |z_{t,i}| \le \lambda_1 \\
+-\frac{z_{t,i} - \text{sgn}(z_{t,i})\lambda_1}{A_{t,i}/\eta_0 - \lambda_2} & \text{otherwise}
+\end{cases}
+$$
+
+Glossing over many technicalities like:
+* Keeping a huge inverse hash map, in order to recover features from hashes;
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: Same as above: lowercase for ";" bullet lists (or uppercase for ".").

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+with $n$ the number of non zero features) even for high dimensional data
+(millions of features). The implementation is pretty generic in the sense that
+---besides the logloss and logit link--- custom loss and link functions can be
+passed as parameters to the optimizer. We implemented all this using Python
+(with patches of C here and there), and we are very happy with our experience
+and results.
+
+Currently, we are well on our way to leverage this technology for the estimation
+of models of market clearing prices and campaign velocity of spend, which are
+paramount to compute the opportunity cost of bidding for a campaign (vs. bidding
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: paramount to compute -> paramount to computing. 
"paramount" needs the "to" preposition after, so "paramount to" expects a noun or verb+ing.

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+for another one) and to establish a bidding schedule with the right pacing for
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: (if "paramount to computing" is changed above) and establishing

File: _src/_posts/2017-01-31-learning-rtb.markdown [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+printing some creative $C$ in the context of the current RTB transaction $T$.
+Why is it so important for our business to have good estimates of these rates?
+Simply because we sell conversions but we pay for impressions. For example,
+suppose one client pays us $\$1$ per install and we estimate the install
+conversion rate to be around $0.1$ (just cheating... real world figures are well
+under a dismaying $0.001$ `:(`). Then, assuming our creative is going to get
+printed, we should expect an average profit of $\$0.1 - c$, where $c$ is the
+expected cost of the impression. If our rate estimates are grossly miscalculated
+we will be losing money by valuating the expected income too low (thus missing
+opportunities with $c < \$0.1$) or too high (thus buying non-opportunities with
```
Comment (GervasioCalderon) [inline _src/_posts/2017-01-31-learning-rtb.markdown]: This sentence is hard to follow: "$c" is written as a variable, but in the example it is supposed to be fixed in some value (say 0.2). The conversion rate is fixed in $0.1$, but it actually depends on the estimations, and should be a variable ("$cvr", for instance). If "valuating the expected income too low " takes place, it's because "$cvr" is too low. But in the example it looks as a fixed 0.1 value.

**Comment by memeplex** [comment]
Muchachos, paren un toque con esto que no es un poema de Oscar Wilde,
además ni sé si todavía lo quieren publicar, y no tengo ganas de hacerle
más de una o dos correcciones como mucho.

On Apr 24, 2017 3:02 PM, "Gervasio Calderón" <notifications@github.com>
wrote:

*@GervasioCalderon* commented on this pull request.
------------------------------

In _src/_posts/2017-01-31-learning-rtb.markdown
<https://github.com/jampp/geeks/pull/27#discussion_r113015383>:

> +conversion.
+
+One main component of our platform is the machine learning subsystem that feeds
+the bidding subsystem with predictive models of conversion rates for different
+events of interest (clicks, installs, opens, etc.). A conversion rate for an
+event $E$ is the conditional probability $p(E|C,T)$ of getting that event by
+printing some creative $C$ in the context of the current RTB transaction $T$.
+Why is it so important for our business to have good estimates of these rates?
+Simply because we sell conversions but we pay for impressions. For example,
+suppose one client pays us $\$1$ per install and we estimate the install
+conversion rate to be around $0.1$ (just cheating... real world
figures are well
+under a dismaying $0.001$ `:(`). Then, assuming our creative is going to get
+printed, we should expect an average profit of $\$0.1 - c$, where $c$ is the
+expected cost of the impression. If our rate estimates are grossly
miscalculated
+we will be losing money by valuating the expected income too low (thus missing
+opportunities with $c < \$0.1$) or too high (thus buying non-opportunities with

This sentence is hard to follow: "$c" is written as a variable, but in the
example it is supposed to be fixed in some value (say 0.2). The conversion
rate is fixed in $0.1$, but it actually depends on the estimations, and
should be a variable ("$cvr", for instance). If "valuating the expected
income too low " takes place, it's because "$cvr" is too low. But in the
example it looks as a fixed 0.1 value.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/jampp/geeks/pull/27#pullrequestreview-34372937>, or mute
the thread
